bin/cron/stop_inactive_adhoc_instances

```diff
@@ -0,0 +1,117 @@
+# set to :stop_instance to stop inactive instances
+EXECUTION_MODE = :warn
+
+# number of http requests received by environment within activity period
+# below which environment is considered to be inactive
+ACTIVITY_THRESHOLD = 1000
+
+# window of time within which activity is measured
+ACTIVITY_PERIOD_DAYS = 7
+
+# no activity is required within the last few days to protect new instances from being terminated
+GRACE_PERIOD_DAYS = 3
+
+# This script stops AWS EC2 instances being used to run adhoc environments if they have not been used recently:
+#
+# 1) for each AWS CloudFormation Stacks with 'adhoc' at the beginning of the name
+#   a) get AWS CloudWatch Metrics to determine if it has been inactive for more than a configurable time period
+#   b) stop instance if it activity is below threshold
+# 2) publish final status to slack cron-job (list of all adhoc environments and which ones were stopped)
+def main
+  ChatClient.message 'cron-daily', 'Beginning to stop inactive adhoc instances'
+  job_status = Array.new
+
+  cloudformation_client = Aws::CloudFormation::Client.new
+  cloudformation_resource = Aws::CloudFormation::Resource.new(client: cloudformation_client)
+  cloudwatch_client = Aws::CloudWatch::Client.new
+
+  cloudformation_resource.stacks.select {|stack| stack.name.starts_with?('adhoc-')}.each do |stack|
+    begin
+      status = StackStatus.new(stack.name, stack.stack_status, stack.creation_time)
+
+      # get Unicorn active HTTP request count CloudWatch metric for the current adhoc stack
+      metric = Aws::CloudWatch::Metric.new(namespace: 'Unicorn', name: 'active', client: cloudwatch_client)
+      unicorn_metric = metric.get_statistics(
+        {
+          dimensions: [
+            {
+              name: 'Environment',
+              value: 'adhoc',
+            },
+            {
+              name: 'Host',
+              value: "#{stack.name}.cdn-code.org",
+            }
+          ],
+          start_time: Time.now - (ACTIVITY_PERIOD_DAYS + GRACE_PERIOD_DAYS).days,
+          end_time: Time.now - GRACE_PERIOD_DAYS.days,
+          period: ACTIVITY_PERIOD_DAYS.days,
+          statistics: ['Sum'],
+          unit: 'Count'
+        }
+      )
+
+      # each datapoint in the metric statistics array contains an attribute called 'sum'
+      total_http_requests = unicorn_metric.datapoints.sum(&:sum) unless unicorn_metric.nil?
+      status[:activity] = total_http_requests
+      instance = Aws::EC2::Instance.new(id: stack.resource('WebServer').physical_resource_id)
+      status[:previous_instance_status] = instance.state.name
+
+      # stop instances with activity below threshold if they are running (state code == 16)
+      if total_http_requests.between?(1, ACTIVITY_THRESHOLD) && instance.state.code == 16
+        puts "Stopping EC2 instance - #{instance.id} - for Stack - #{stack.name} - due to inactivity"
+        stop_result = instance.stop(
```

Comment:
Does the CF stack also need to be deleted after the instance is stopped? Or is the idea just to stop the instances and not fully delete everything?

Reply:
The intention with this job is to just warn that an EC2 Instance resource is not being used and/or to stop the Instance.  We would still manually invoke `rake adhoc:stop`  to tear down the Stack and all of its resources.